Add @username suggestions to notification and reader full comment view.

hafizrahman opened the pull request:
Fixes #10328 

To test:

### On notification area
1. Go to Notifications tab.
2. Select Comments tab.
3. Find a comment to a site of yours that's on WordPress.com, tap it.
4. Tap Reply field.
5. Enter @ followed by username of site user or the commenter, assuming they're WordPress.com accounts.
6. Notice username suggestions are shown (this is already working prior to this PR, but just to check).
7. Delete text from Reply field.
8. Tap Expand action in Reply field.
9. Enter @ followed by username of site user or the commenter, assuming they're WordPress.com accounts. 
10. Notice username suggestions are now shown.

### On reader area
1. Go to Reader tab.
2. Find a WordPress.com site post, probably easiest if it's your own test site.
3. Tap the comment icon on that post.
4. Tap Reply field.
5. Enter @ followed by username of site user or the commenter, assuming they're WordPress.com accounts.
6. Notice username suggestions are shown (this is already working prior to this PR, but just to check).
7. Delete text from Reply field.
8. Tap Expand action in Reply field.
9. Enter @ followed by username of site user or the commenter, assuming they're WordPress.com accounts. 
10. Notice username suggestions are now shown.

PR submission checklist:

- [x] I have considered adding unit tests where possible.

- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

@shiki Question about that last checkbox above, is this change something that warrants being added to RELEASE-NOTES.txt?

shiki commented on WordPress/src/main/java/org/wordpress/android/ui/CommentFullScreenDialogFragment.kt:
```diff
@@ -95,12 +118,14 @@ class CommentFullScreenDialogFragment : Fragment(), CollapseFullScreenDialogCont
+        private const val EXTRA_SITE_MODEL = "EXTRA_SITE_MODEL"
 
-        fun newBundle(reply: String, selectionStart: Int, selectionEnd: Int): Bundle {
+        fun newBundle(reply: String, selectionStart: Int, selectionEnd: Int, mSite: SiteModel): Bundle {
```
We now prefer to pass the `id` instead of the full object. Passing an object in a `Bundle` can easily lead to crashes [like this](https://github.com/wordpress-mobile/WordPress-Android/pull/10240). `Bundle` has a limit:

> When sending data via an intent, you should be careful to limit the data size to a few KB. Sending too much data can cause the system to throw a TransactionTooLargeException exception.

_from: https://developer.android.com/guide/components/activities/parcelables-and-bundles_

It's generally good practice to use scalar values whenever possible. 

In this case, I think we can pass `SITE_ID` and load the `SiteModel` instead. 🙂

shiki commented on WordPress/src/main/java/org/wordpress/android/ui/CommentFullScreenDialogFragment.kt:
```diff
@@ -55,11 +59,30 @@ class CommentFullScreenDialogFragment : Fragment(), CollapseFullScreenDialogCont
+        if (!isAdded || mSite == null || !SiteUtils.isAccessedViaWPComRest(mSite)) {
+            return
+        }
+        val mSuggestionServiceConnectionManager = SuggestionServiceConnectionManager(
```
The `m` prefix is only used for Java instance variables. For Kotlin, we don't use it. 🙂

shiki commented on WordPress/src/main/java/org/wordpress/android/ui/CommentFullScreenDialogFragment.kt:
```diff
@@ -55,11 +59,30 @@ class CommentFullScreenDialogFragment : Fragment(), CollapseFullScreenDialogCont
         return layout
     }
 
+    private fun setupSuggestionServiceAndAdapter(mSite: SiteModel, reply: SuggestionAutoCompleteText) {
```
Is there a reason why you passed `reply` here and not access `this.reply` directly?

hafizrahman replied:
I tried a bunch of things to refer to the view that didn't work, and things got weird and decided to just pass it as a parameter. I changed this up and can confirm that using `this.reply` works, though, thanks!.

shiki commented on WordPress/src/main/java/org/wordpress/android/ui/CommentFullScreenDialogFragment.kt:
```diff
@@ -55,11 +59,30 @@ class CommentFullScreenDialogFragment : Fragment(), CollapseFullScreenDialogCont
     }
 
+    private fun setupSuggestionServiceAndAdapter(mSite: SiteModel, reply: SuggestionAutoCompleteText) {
+        if (!isAdded || mSite == null || !SiteUtils.isAccessedViaWPComRest(mSite)) {
```
Android Studio shows a warning/info for `mSite == null`  since `mSite` will never be `null`. It probably will after the change to `SITE_ID` though (please see my comment below).

hafizrahman commented:
> It looks like there is a bug in the suggestions view.

@shiki Thanks for reviewing things! As for this part, it seems like the dropdown generated by `AutoCompleteTextView` can only be set on a certain x,y location, or [be anchored to a certain view](https://developer.android.com/reference/android/widget/AutoCompleteTextView.html#setDropDownAnchor(int)). As far as I can tell, its position can't be set dynamically depending on the current triggering string cursor's location (so it can't show up right above / bottom the `@...` text, for example).

[Here's the Slack Overflow link](https://stackoverflow.com/questions/26401412/android-autocompletetextview-dropdown-position) that was my starting point for the research.

In this case I think the current positioning at the top is already the ideal compromise, it will remain viewable and selectable regardless of the current text cursor position.

hafizrahman commented:
Found a bug where if there's multiple suggestions shown in the dropdown, it's only displaying one and the rest is kind of hard to see:

Showing only first result:
<img width="430" alt="Screen Shot 2019-11-16 at 15 55 44" src="https://user-images.githubusercontent.com/266376/68990777-c52fb000-0889-11ea-8c68-885251837499.png">
<img width="444" alt="Screen Shot 2019-11-16 at 15 56 01" src="https://user-images.githubusercontent.com/266376/68990779-c6f97380-0889-11ea-9d78-8975257b3da5.png">

Second result only shown after scrolling:

jkmassel commented:
Howdy folks! I'm freezing `13.7` today, so this is being bumped to `13.8`. If you'd like it to go out as part of `13.7`, please target `release/13.7` for this PR, and DM me once it's merged – I'll be happy to cut a new beta release!

shiki commented:
## Design Review Request

Hi Design Team! 👋 I was wondering you could give some inputs on this. Currently, the dropdown will always be shown at the top even if the cursor is somewhere in the bottom (https://github.com/wordpress-mobile/WordPress-Android/pull/10781#pullrequestreview-317341197). Is there an existing design within the app that we can follow? 

I spent some time to look at whether we can follow the cursor position. It seems quite challenging. 🤔

osullivanchris commented:
@shiki Hey thanks for the ping. Strange, I was thinking I would really like to have it next to the cursor. However I think if its fixed to something, I think fixed to the keyboard can work really well.

In the left of this sketch, the suggestion box appears above the keyboard. There is an item cut off to indicate its scrollable.

On the right: I know that we can't make it dynamic depending on cursor position. But can we make it dynamic depending on the total amount of content? So that more of the list is visible when the space is available and its always close to what you're writing?

Last note - in the sketch I have drawn this full width as I think it can work nicely with the keyboard and feel more natural to scroll within. Not sure if that is easy or problematic with this component.

Keep me in the loop!

![IMG_1546](https://user-images.githubusercontent.com/50576199/69799647-9c81b180-11cb-11ea-947b-225a6ec5101c.JPG)

shiki commented:
Thank you, @osullivanchris! Making it full width and fixing it to the keyboard sounds doable. I've seen other apps doing something similar. Making the height dynamic could be a nice-to-have but we can attempt it.

jkmassel commented:
We're freezing `13.8` today, so I'm moving this forward to `13.9`. If you'd like it released as part of `13.8`, please adjust the target from `develop` to `release/13.8`, and let me know once this PR has been merged – I'll be happy to ship a new release candidate!

hafizrahman commented:
@shiki I added some layout fixes to make the suggestion drop down stick to the bottom. This also fixed the bug [I mentioned above](https://github.com/wordpress-mobile/WordPress-Android/pull/10781#issuecomment-554618532) as far as I can check.

Some screenshots:

1. When the mention is at the top:
![Screenshot_1576826851](https://user-images.githubusercontent.com/266376/71238793-9351b380-2337-11ea-89d9-f70683ddd32d.png)

2. When the mention is in the middle of the comment and showing multiple mentions:
![Screenshot_1576826912](https://user-images.githubusercontent.com/266376/71238796-964ca400-2337-11ea-860e-687f4fa61f89.png)

3. When the mention is in the middle of the comment and there's only one result:
![Screenshot_1576826922](https://user-images.githubusercontent.com/266376/71238802-98aefe00-2337-11ea-800a-bd854203443f.png)

osullivanchris commented:
Thanks @hafizrahman  looks much better! Nice job

shiki commented on WordPress/src/main/res/layout/comment_dialog_fragment.xml:
```diff
@@ -1,27 +1,41 @@
+            app:layout_constraintTop_toTopOf="parent"
+            android:dropDownAnchor="@+id/dummy_hidden_view" />
+
+        <View
```
That's really smart. 👍
